src/lib/updating-element.ts

```diff
@@ -0,0 +1,503 @@
+  ensurePropertyStorage(ctor);
+  ctor._classProperties[name] = options || {};
+  makeProperty(name, proto);
+};
+
+
+/**
+ * AttributeProcessor which configures properties which should reflect to and
+ * from boolean attributes. If the attribute exists, the property becomes true;
+ * and if the property value is truthy, the attribute is set to an empty string.
+ */
+export const BooleanAttribute: AttributeProcessor = {
+  fromAttribute: (value: string) => value !== null,
+  toAttribute: (value: string) => value ? '' : null
+};
+
+/**
+ * Change function that returns true if `value` is different from `oldValue`.
+ * This method is used as the default for a property's `shouldChange` function.
+ */
+// tslint:disable-next-line no-any
+export const identity = (value: any, old: any) => {
+  // This ensures (old==NaN, value==NaN) always returns false
+  return old !== value && (old === old || value === value);
+};
+
+enum ValidationState {
+  Disabled = 0,
+  Valid,
+  Invalid
+}
+
+/**
+ * Base element class which manages element properties and attributes. When
+ * properties change, the `update` method is asynchronously called. This method
+ * should be supplied by subclassers to render updates as desired.
+ */
+export abstract class UpdatingElement extends HTMLElement {
+
+  /**
+   * Maps attribute names to properties; for example `foobar` attribute
+   * to `fooBar` property.
+   */
+  private static _attributeToPropertyMap: AttributeMap = {};
+  /**
+   * Marks class as having finished creating properties.
+   */
+  private static _finalized = true;
+  /**
+   * Memoized result of computed observedAttributes.
+   */
+  private static _observedAttributes: string[]|undefined;
+
+  // TODO(sorvell): intended to be private but called by decorator.
+  /**
+   * Memoized list of all class properties, including any superclass properties.
+   */
+  static _classProperties: Properties = {};
+
+  static properties: Properties = {};
+
+  /**
+   * Returns a list of attributes corresponding to the registered properties.
+   */
+  static get observedAttributes() {
+    if (!this.hasOwnProperty('_observedAttributes')) {
+      // note: piggy backing on this to ensure we're _finalized.
+      this._finalize();
+      this._observedAttributes = [];
+      for (const p in this._classProperties) {
+        const attr = this.attributeNameForProperty(p);
+        if (attr !== undefined) {
+          this._attributeToPropertyMap[attr] = p;
+          this._observedAttributes.push(attr);
+        }
+      }
+    }
+    return this._observedAttributes;
+  }
+
+  /**
+   * Creates property accessors for registered properties and ensures
+   * any superclasses are also finalized.
+   */
+  private static _finalize() {
+    if (this.hasOwnProperty('_finalized') && this._finalized) {
+      return;
+    }
+    // finalize any superclasses
+    const superCtor = Object.getPrototypeOf(this);
+    if (superCtor.prototype instanceof UpdatingElement) {
```

Comment:
Rather than an instanceof check, can we rely on the presence of the `_finalized` method? This would make subclassing work across multiple versions of this package.

Reply:
Done